### PR TITLE
Create Benchmarks Package for State Transition

### PR DESCRIPTION
**Pull request by 0xKiwi**
Part of #3668, this PR adds the ability to benchmark the beacon state transition for a large amount of validators.

There are two packages, with a readme to explain their usage.: 
`benchmark_files`: A binary to generate the `.ssz` files that are marshalled blocks and beacon states.
`benchmarks`: For the benchmark functionality itself that uses the marshalled blocks and beacon states, to prevent duplicated computation.

Note, while `benchmark_files` is a binary, since it has to add files to the local directory, it's better to use `go run` to execute it rather than `bazel run`.

### CODE REVIEW

File: beacon-chain/core/blocks/benchmarks_test.go [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+	for i := 0; i < b.N; i++ {
+		fmt.Println(i)
+		if _, err := state.ExecuteStateTransition(context.Background(), cleanStates[i], block); err != nil {
+			b.Fatal(err)
+		}
+	}
+}
+
+func BenchmarkHashTreeRoot_65536Validators(b *testing.B) {
+	slotsPerEpoch := params.BeaconConfig().SlotsPerEpoch
```
Comment (nisdas) [inline beacon-chain/core/blocks/benchmarks_test.go]: this whole chunk of setting up to run a transition can be abstracted into a separate method.

**Comment by 0xKiwi** [comment]
The SSZ files (at least the genesis state one) is needed because of how long it takes to generate a beacon state of 66536 validators.

**Comment by nisdas** [comment]
Ah yeap, forgot about those. Its fine in that case then

**Comment by rauljordan** [comment]
Any update on this @0xKiwi ?

**Comment by 0xKiwi** [comment]
> Any update on this @0xKiwi ?

Just synced up with Preston on this yesterday, main blocker on this is the SSZ error with HashTreeRoot'ing 128 attestations in a block. In the meantime I'll clean this up and just have it run with 127 attestations so we can check it in.

**Comment by rauljordan** [comment]
We still have the ssz error for 128 attestations?

**Comment by 0xKiwi** [comment]
Seems like we don't @rauljordan, can't replicate.

File: beacon-chain/core/blocks/benchmarks/benchmarks_test.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	"github.com/prysmaticlabs/go-ssz"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/featureconfig"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/testutil"
+)
+
+var validatorCount = 65536
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/benchmarks/benchmarks_test.go]: Please add a comment to explain this value. Is it mainnet?

File: beacon-chain/core/helpers/attestation.go [hunk 3]
```diff
@@ -144,7 +143,6 @@ func IsAggregator(state *pb.BeaconState, slot uint64, index uint64, slotSig *bls
 	if len(committee)/int(params.BeaconConfig().TargetAggregatorsPerCommittee) > 1 {
 		modulo = uint64(len(committee)) / params.BeaconConfig().TargetAggregatorsPerCommittee
 	}
-	fmt.Println(modulo)
```
Comment (prestonvanloon) [inline beacon-chain/core/helpers/attestation.go]: Nice

File: beacon-chain/core/blocks/benchmarks/benchmark_files/generate_bench_files.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/interop"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/testutil"
+)
+
+func main() {
+	log.Println("generating genesis state")
```
Comment (rauljordan) [inline beacon-chain/core/blocks/benchmarks/benchmark_files/generate_bench_files.go]:
```suggestion
	log.Println("Generating genesis state")
```

File: beacon-chain/core/blocks/benchmarks/benchmark_files/generate_bench_files.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	// Generating this for the 2 following states.
+	if err := generateGenesisBeaconState(); err != nil {
+		log.Fatal(err)
+	}
+	log.Println("generating full block and state after 1 skipped epoch")
```
Comment (rauljordan) [inline beacon-chain/core/blocks/benchmarks/benchmark_files/generate_bench_files.go]:
```suggestion
	log.Println("Generating full block and state after 1 skipped epoch")
```

File: beacon-chain/core/blocks/benchmarks/benchmark_files/generate_bench_files.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	if err != nil {
+		return err
+	}
+
+	conf := &testutil.BlockGenConfig{
+		MaxAttestations: 0,
+		Signatures:      true,
+	}
+
+	// Small offset for the beacon state so we dont process a block on an epoch
```
Comment (rauljordan) [inline beacon-chain/core/blocks/benchmarks/benchmark_files/generate_bench_files.go]:
```suggestion
	// Small offset for the beacon state so we dont process a block on an epoch.
```

File: beacon-chain/core/blocks/benchmarks/config.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
+package benchmarks
+
+import (
+	"fmt"
+
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// ValidatorCount is for declaring how many validators the benchmarks will be
```
Comment (rauljordan) [inline beacon-chain/core/blocks/benchmarks/config.go]: Why do these need to be exported @0xKiwi?

Comment (rauljordan) [reply beacon-chain/core/blocks/benchmarks/config.go]: Any chance we could avoid this if we put this config file next to the place it's needed?

Comment (0xKiwi) [reply beacon-chain/core/blocks/benchmarks/config.go]: It's needed in the `benchmarks` package and the `benchmark_files` package. They needed to be exported so the `benchmark_files` package can access what they are.

Comment (rauljordan) [reply beacon-chain/core/blocks/benchmarks/config.go]: Got it, that makes sense

**Pull request merged**